### PR TITLE
osd: not flush or evict object which is in object_contexts

### PR DESCRIPTION
**Pull request by XinzeChi**
I think object in object_contexts is hot object.

Signed-off-by: Xinze Chi <xmdxcxz@gmail.com>

### CODE REVIEW

**Comment by liewegas** [comment]
This looks like a good strategy.  My only concern is if the obc cache size is larger than the tiering limits.  Perhaps we should check that the per-pg limit (max_dirty_objects or whatever for the pool) is higher than the obc cache size (or cache size * 2 maybe for good measure) before using the cache as an indicator.  Otherwise we could prevent flushing.  My guess is as is this will break various tests that set the max_dirty_objects to very low numbers of ensure the agent is working correctly.

**Comment by XinzeChi** [comment]
ping @liewegas

File: src/osd/ReplicatedPG.cc [hunk 1]
```diff
@@ -10477,6 +10477,24 @@ void ReplicatedPG::agent_clear()
   agent_state.reset(NULL);
 }
 
+bool ReplicatedPG::obccache_indicate_flush_evict(hobject_t& o)
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: obccache_suggests_hot() ?

File: src/osd/ReplicatedPG.cc [hunk 2]
```diff
@@ -10477,6 +10477,24 @@ void ReplicatedPG::agent_clear()
...
+{
+  if (!object_contexts.lookup(o)) {
+    return false;
+  }
+  uint64_t divisor = pool.info.get_pg_num_divisor(info.pgid.pgid);
+  assert(divisor > 0);
+  uint64_t max_dirty_objects = (uint64_t)(pool.info.target_max_objects *
+    (float)pool.info.cache_target_dirty_ratio_micro / 1000000.0);
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: * 1000000 / ..._micro (and avoid the floating point arithmetic?).  should be no danger of overflow here.

File: src/osd/ReplicatedPG.cc [hunk 3]
```diff
@@ -10477,6 +10477,24 @@ void ReplicatedPG::agent_clear()
...
+  max_dirty_objects = max_dirty_objects / divisor;
+  dout(20) << __func__ << " max_dirty_objects " << max_dirty_objects
+           << " object_contexts size " << object_contexts.size() << dendl;
+  if ( max_dirty_objects > object_contexts.size() * 2) {
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: s/if ( /if (/

**Comment by liewegas** [comment]
just cosmetic issues, otherwise looks good!

**Comment by XinzeChi** [comment]
ping @liewegas

**Comment by liewegas** [comment]
Reviewed-by:

**Comment by athanatos** [comment]
I don't think this is a good strategy.  There are a bunch of background processes which polute the obc cache.  It would be much better to maintain an in-memory last-N list of client object accesses and be done with it.

**Comment by athanatos** [comment]
It seems like at best, this will duplicate the information in the hitsets.  I'm closing it for now unless someone strongly disagrees.

**Comment by liewegas** [comment]
Chatted with @athanatos and agree that this doesn't make sense for flush, since the object will be in the hitset in that case (unless we're missing something)?

I think we *do* want to bit here that does the warmup time period though so that the similar NOCACHE->WONTNEED heuristic can use it.

so, we want https://github.com/ceph/ceph/pull/4392 and https://github.com/ceph/ceph/pull/3868, and i think the part of this patch that does

+bool ReplicatedPG::obccache_suggests_hot(hobject_t& o)

?

@XinzeChi do you mind combining these patches all into a single pull request where we can look at them together?  Thanks!

**Comment by XinzeChi** [comment]
@liewegas, I think the obc_cache is not same as hitset.
During the same hit_set_period, osd do not distinguish between them. And if object is accessed many times(read or write) by client during the hit_set_period, osd do not record it in hitset. But obc_cache could distinguish.
So I agree with you combining 4392 and #3868 together. I would to that.

**Comment by XinzeChi** [comment]
@liewegas , @majianpeng , I have combined the #4392 and #3868 together at https://github.com/ceph/ceph/pull/4392.

**Pull request closed without merging**